util/confutil/config.go

```diff
@@ -1,39 +1,140 @@
 
 import (
+	"crypto/rand"
+	"encoding/hex"
 	"os"
-	"path"
 	"path/filepath"
+	"strings"
+	"sync"
 
 	"github.com/docker/cli/cli/command"
+	"github.com/docker/docker/pkg/ioutils"
 	"github.com/pelletier/go-toml"
 	"github.com/pkg/errors"
-	"github.com/sirupsen/logrus"
 )
 
-// ConfigDir will look for correct configuration store path;
-// if `$BUILDX_CONFIG` is set - use it, otherwise use parent directory
-// of Docker config file (i.e. `${DOCKER_CONFIG}/buildx`)
-func ConfigDir(dockerCli command.Cli) string {
-	if buildxConfig := os.Getenv("BUILDX_CONFIG"); buildxConfig != "" {
-		logrus.Debugf("using config store %q based in \"$BUILDX_CONFIG\" environment variable", buildxConfig)
-		return buildxConfig
+var (
+	sudoerUID = -1
+	sudoerGID = -1
+)
+
+type Config struct {
+	dir string
+}
+
+type ConfigOption func(*configOptions)
+
+type configOptions struct {
+	dir string
+}
+
+func WithDir(dir string) ConfigOption {
+	return func(o *configOptions) {
+		o.dir = dir
 	}
+}
 
-	buildxConfig := filepath.Join(filepath.Dir(dockerCli.ConfigFile().Filename), "buildx")
-	logrus.Debugf("using default config store %q", buildxConfig)
-	return buildxConfig
+func NewConfig(dockerCli command.Cli, opts ...ConfigOption) *Config {
+	co := configOptions{}
+	for _, opt := range opts {
+		opt(&co)
+	}
+	configDir := co.dir
+	if configDir == "" {
+		configDir = os.Getenv("BUILDX_CONFIG")
+		if configDir == "" {
+			configDir = filepath.Join(filepath.Dir(dockerCli.ConfigFile().Filename), "buildx")
+		}
+	}
+	return &Config{
+		dir: configDir,
+	}
 }
 
-// DefaultConfigFile returns the default BuildKit configuration file path
-func DefaultConfigFile(dockerCli command.Cli) (string, bool) {
-	f := path.Join(ConfigDir(dockerCli), "buildkitd.default.toml")
+// Dir will look for correct configuration store path;
+// if `$BUILDX_CONFIG` is set - use it, otherwise use parent directory
+// of Docker config file (i.e. `${DOCKER_CONFIG}/buildx`)
+func (c *Config) Dir() string {
+	return c.dir
+}
+
+// BuildKitConfigFile returns the default BuildKit configuration file path
+func (c *Config) BuildKitConfigFile() (string, bool) {
+	f := filepath.Join(c.dir, "buildkitd.default.toml")
 	if _, err := os.Stat(f); err == nil {
 		return f, true
 	}
 	return "", false
 }
 
+// MkdirAll creates a directory and all necessary parents within the config dir
+func (c *Config) MkdirAll(dir string, perm os.FileMode) error {
+	d := filepath.Join(c.dir, dir)
+	if err := os.MkdirAll(d, perm); err != nil {
+		return err
+	}
+	if sudoerUID != -1 && sudoerGID != -1 {
+		// apply chown to each directory level
+		parts := strings.Split(dir, string(filepath.Separator))
+		for i := range parts {
+			if err := os.Chown(filepath.Join(c.dir, filepath.Join(parts[:i+1]...)), sudoerUID, sudoerGID); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// WriteFile writes data to a file within the config dir
+func (c *Config) WriteFile(filename string, data []byte, perm os.FileMode) error {
```

Comment:
Why do we need non-atomic write?

Reply:
This is the current behavior for https://github.com/docker/buildx/blob/9f0ebd2643482722e54eae975df4193df5b456f1/util/confutil/node.go#L23

But we can switch to atomic.

Reply:
Yes, this can be atomic.